src/map/utils/charutils.cpp

```diff
@@ -4404,6 +4409,11 @@ namespace charutils
         else
             exp = exp + bonus;
 
+        if (hasKeyItem(PChar, 2884)) // Rhapsody in White
```

Comment:
I'm unfamiliar with AddExpBonus, and this may be in "premature optimization" territory, but I'm guessing AddExpBonus is called every time upon a mob kill, and there will eventually be 9 ROV KIs.

I'm wondering that, since these KI effects are permanent upon being obtained (they don't wear off like Dedication would), if instead of 9 `if has KI` checks and combining them upon each exp event, if instead the ROV bonus should be calculated and stored as a `MOD_ROV_EXP` or something on zone/character load/whatever, and then in AddExpBonus just `exp *= MOD_ROV_EXP`

(Same thing for skill-ups)

Reply:
Like I said in discord yesterday, probably store a total in either a mod or a variable (you will be checking those KI even if it were a mod) and then multiply by that value at the end just before exp dispersement.